src/log_rate.c

```diff
@@ -0,0 +1,110 @@
+static int64_t subtract_timespecs_nano(const struct timespec* first, const struct timespec* second);
+static void write_io_bufs(stdpipe_t pipe, char* buf, ssize_t count);
+static void sleep_for_the_rest_of_this_period();
+static void start_new_period();
+
+void log_rate_init() {
+	start_new_period();
+}
+
+bool log_rate_write_to_logs(stdpipe_t pipe, char *buf, ssize_t num_read) {
+	char* buf_start = buf;
+	ssize_t bytes_remaining = num_read;
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	int64_t diff_nano = subtract_timespecs_nano(&now, &start_of_this_period);
+	if (diff_nano < SECS_PER_PERIOD * BILLION) {
+		ssize_t bytes_we_can_write = BYTES_PER_PERIOD - bytes_written_this_period;
+		if (num_read <= bytes_we_can_write) {
+			write_io_bufs(pipe, buf_start, num_read);
+			bytes_written_this_period += num_read;
+			return true;
+		} else {
+			write_io_bufs(pipe, buf_start, bytes_we_can_write);
+			bytes_written_this_period += bytes_we_can_write;
+			buf_start += bytes_we_can_write;
+			bytes_remaining = num_read - bytes_we_can_write;
+			sleep_for_the_rest_of_this_period();
+			start_new_period();
+		}
+	} else {
+		start_new_period();
+	}
+
+	ssize_t chunks = bytes_remaining / BYTES_PER_PERIOD;
+	ssize_t remainder = bytes_remaining % BYTES_PER_PERIOD;
+
+	for (ssize_t i = 0; i < chunks; ++i) {
+		write_io_bufs(pipe, buf_start + i * BYTES_PER_PERIOD, BYTES_PER_PERIOD);
+		sleep_for_the_rest_of_this_period();
+		start_new_period();
+	}
+	if (remainder != 0) {
+	    if (bytes_written_this_period + remainder > BYTES_PER_PERIOD) {
+		    sleep_for_the_rest_of_this_period();
+		    start_new_period();
+	    }
+	    write_io_bufs(pipe, buf_start + (chunks * BYTES_PER_PERIOD), remainder);
+	    bytes_written_this_period += remainder;
+	}
+	return true;
+}
+
+struct timespec subtract_timespecs(const struct timespec* first, const struct timespec* second) {
+	int64_t diff_nanoseconds = subtract_timespecs_nano(first, second);
+	struct timespec ret = {diff_nanoseconds / BILLION, diff_nanoseconds % BILLION};
+	return ret;
+}
+
+int64_t subtract_timespecs_nano(const struct timespec* first, const struct timespec* second) {
+	return (first->tv_sec - second->tv_sec) * BILLION + first->tv_nsec - second->tv_nsec;
+}
+
+void write_io_bufs(stdpipe_t pipe, char* buf, ssize_t count) {
+	ssize_t chunks = count / IO_BUF_SIZE;
+	ssize_t remainder = count % IO_BUF_SIZE;
+	for (int i = 0; i < chunks; ++i) {
+		write_to_logs(pipe, buf + (i * IO_BUF_SIZE), IO_BUF_SIZE);
+	}
+	write_to_logs(pipe, buf + count - remainder, remainder);
+}
+
+void sleep_for_the_rest_of_this_period() {
+	int ret;
+	struct timespec now, sleep, diff;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	diff = subtract_timespecs(&now, &start_of_this_period);
+	sleep = subtract_timespecs(&secs_per_period, &diff);
+	if (sleep.tv_sec < 0 || sleep.tv_nsec < 0)
+		nexit("negative sleep");
+	do {
+		ret = nanosleep(&sleep, &sleep);
+	} while (ret == -1 && errno == EINTR);
```

Comment:
Not sure you want a loop that doesn't check the time change.  Shouldn't we be checking to see if the absolute time has reached or passed the calculated absolute time for the sleep?

Reply:
My reading of nanosleep(2) is that it sleeps for relative time intervals and when interrupted by a signal, it will store the remaining time to sleep in the timespec pointed to by the second argument, which we will use on the next pass through the loop as the relative time to sleep.

And I think I'm not doing the right thing for the negative sleep - if it's negative, it means we're in a new period and we should just return without sleeping. We start a new period after every sleep anyway.